design/design-doc-smaller-providers.md

```diff
@@ -0,0 +1,429 @@
+
+* No change to the compute resources consumed by Terraform CLI invocations.
+* Fewer compute resources consumed by Crossplane provider processes.
+* Significantly more compute resources consumed by Terraform provider processes.
+
+We expect CPU and memory used by Terraform CLI invocations to remain the same
+for the same number of MRs. This is because the upjet providers invoke the
+Terraform CLI 2-3 times per reconcile. The number of reconciles won’t change -
+they’ll just be spread across multiple Provider pods.
+
+We expect fewer compute resources (in particular less memory) will be consumed
+by Crossplane provider processes because we’ll be running fewer controllers
+overall. Today provider-aws, at idle, must start almost 900 controllers. One for
+each supported kind of managed resource. Each controller spins up a few
+goroutines and subscribes to API server updates for the type of MR it
+reconciles. By breaking up providers significantly fewer controllers would be
+running, using resources at idle.
+
+We expect significantly more compute resources to be consumed by Terraform
+provider processes because there will need to be more of them. It’s safe to
+assume each deployment of provider-aws becomes ~7 smaller deployments on
+average. For a single ProviderConfig, each of these smaller deployments would
+incur the ~300MB of memory required to run a Terraform provider process. This
+doubles as we increase to two ProviderConfigs, etc.
+
+The increased compute resources consumed by needing to run Terraform providers
+is a serious concern. I’m not considering it a blocker for this proposal because
+we feel there is a lot of room for improvement in the medium to long term. In
+addition to the current steps being investigated to improve performance, we
+could consider:
+
+* Replacing the most widely-used MR controllers with native implementations that
+  use dramatically fewer resources.
+* Cutting out Terraform CLI invocations by talking directly to the Terraform
+  provider process.
+* Cutting out both Terraform CLI invocations and Terraform provider processes by
+  importing the Terraform provider code directly. Terraform makes this hard to
+  do, but it’s possible. It appears to be what Config Connector does.
+
+## Alternatives Considered
+
+I considered the following alternatives before making this proposal.
+
+### CRD Filtering
+
+The gist of this proposal is to update the existing Provider.pkg.crossplane.io
+API to include a configurable allow list of enabled CRDs. As far as I can tell
+this is the only feasible alternative to breaking up the largest Providers.
+
+If we took this approach I would propose that:
+
+* We teach the Package Manager that only some Providers support filtering. Such
+  Providers would have something like `pkg.crossplane.io/filtering: enabled` in
+  their `crossplane.yaml`. This is necessary to avoid filtering being silently
+  ineffectual when not supported by a Provider.
```

Comment:
I don't love it. I could agree that it's "good enough", but it doesn't seem like an ideal UX that if you configure your `Provider.pkg` in a certain way (that has nothing obvious to do with CLI flags) you risk your provider failing to install with an "unknown CLI flag" error.

Reply:
> you risk your provider failing to install with an "unknown CLI flag" error.

Given this is a recoverable error with no harm by removing `enabledAPIs` from `Provider.pkg`, I don't see this as a big risk.

I don't love it too, but given that it is some transitive behavior until providers support filtering I still believe we could live with it.

Reply:
I touch on this now in the design. If we did go the filtering route I think we should do this - it shouldn't be that hard to do and it would provide a better UX.